pkg/kube/controllers/extendedjob/job_create_secret.go

```diff
@@ -33,106 +32,137 @@ var (
+	pod, err := po.clientSet.CoreV1().Pods(po.namespace).Get(po.podName, metav1.GetOptions{})
 	if err != nil {
-		return errors.Wrapf(err, "failed to fetch pod %s", podName)
+		return errors.Wrapf(err, "failed to fetch pod %s", po.podName)
 	}
 
 	// Fetch the exjob
 	exjobName := pod.GetLabels()[LabelEjobPod]
 
-	exjobClient := versionedClientSet.ExtendedjobV1alpha1().ExtendedJobs(namespace)
+	exjobClient := po.versionedClientSet.ExtendedjobV1alpha1().ExtendedJobs(po.namespace)
 	exjob, err := exjobClient.Get(exjobName, metav1.GetOptions{})
 	if err != nil {
 		return errors.Wrapf(err, "failed to fetch exjob")
 	}
 
-	// convert output if needed
-	if exjob.Spec.Output != nil {
-		err = convertOutputToSecret(pod, exjob.Spec.Output.NamePrefix, namespace, clientSet, exjob)
+	// Persist output if needed
+	if !reflect.DeepEqual(ejv1.Output{}, exjob.Spec.Output) && exjob.Spec.Output != nil {
+		err = po.ConvertOutputToSecretPod(pod, exjob)
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func convertOutputToSecret(pod *corev1.Pod, namePrefix string, namespace string, clientSet *kubernetes.Clientset, exjob *ejv1.ExtendedJob) error {
-	fileNotifyChannel := make(chan int)
-	errorChannel := make(chan error)
+// ConvertOutputToSecretPod starts goroutine for converting each container
+// output into a secret.
+func (po *PersistOutputInterface) ConvertOutputToSecretPod(pod *corev1.Pod, exjob *ejv1.ExtendedJob) error {
 
-	// Loop over containers and create secrets for each output file for container
-	for containerIndex, container := range pod.Spec.Containers {
+	errorContainerChannel := make(chan error)
 
+	// Loop over containers and create go routine
+	for containerIndex, container := range pod.Spec.Containers {
 		if container.Name == "output-persist" {
 			continue
 		}
-
-		filePath := filepath.Join("/mnt/quarks/", container.Name, "output.json")
-
-		// Go routine to wait for the file to be created
-		go waitForFile(containerIndex, filePath, fileNotifyChannel, errorChannel, container.Name)
+		go po.ConvertOutputToSecretContainer(containerIndex, container, exjob, errorContainerChannel)
 	}
 
-	// wait for all the go routines
+	// wait for all container go routines
 	for i := 0; i < len(pod.Spec.Containers)-1; i++ {
-		select {
-		case containerIndex := <-fileNotifyChannel:
-			outputContainer := pod.Spec.Containers[containerIndex]
-			err := createOutputSecret(outputContainer, namePrefix, namespace, pod.Name, clientSet, exjob)
+		err := <-errorContainerChannel
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ConvertOutputToSecretContainer converts json output file
+// of the specified container into a secret
+func (po *PersistOutputInterface) ConvertOutputToSecretContainer(containerIndex int, container corev1.Container, exjob *ejv1.ExtendedJob, errorContainerChannel chan<- error) {
+
+	filePath := filepath.Join(po.outputFilePathPrefix, container.Name, "output.json")
```

Comment:
Do you need the `container.Name` to be inside the path?

Reply:
Yes, as this code is run the side container and volume mount paths are like /mnt/quarks/c1/output.json, /mnt/quarks/c2/output.json.